velox/exec/Exchange.cpp

```diff
@@ -181,6 +206,29 @@ std::unique_ptr<ExchangeSource> createLocalExchangeSource(
 
 } // namespace
 
+uint64_t ExchangeQueue::expectedSerializedPageSize() const {
+  // Initially, max size / number of sources. Later, average of received sizes.
+  constexpr int32_t kMinExpectedSize = PartitionedOutput::kMinDestinationSize;
+  if (!numSources_) {
+    return kMinExpectedSize;
+  }
+  auto numSourcesGuess =
+      noMoreSources_ ? numSources_ : std::max(numSources_, 100);
+  return std::max<int64_t>(
+      kMinExpectedSize,
+      (maxBytes_ + receivedBytes_) / (receivedPages_ + numSourcesGuess));
+}
+
+bool ExchangeQueue::requestIfDue(
```

Comment:
I suggest to modify this method to take no arguments and just return a boolean indicating whether the caller should keep fetching more data or not.

Reply:
replySequence is an implementation detail of the ExchangeSource that is currently leaking out.